src/coreclr/jit/fgprofilesynthesis.cpp

```diff
@@ -976,11 +1004,25 @@ void ProfileSynthesis::ComputeBlockWeights()
 {
     JITDUMP("Computing block weights\n");
 
+    bool useSolver = true;
+
+#ifdef DEBUG
+    useSolver = JitConfig.JitSynthesisUseSolver() > 0;
+#endif
+
+    if (useSolver)
+    {
+        GaussSeidelSolver();
+        return;
+    }
+
     for (unsigned i = m_dfsTree->GetPostOrderCount(); i != 0; i--)
     {
         BasicBlock* block = m_dfsTree->GetPostOrder(i - 1);
         ComputeBlockWeight(block);
     }
+
+    m_approximate = (m_cappedCyclicProbabilities) || (m_improperLoopHeaders > 0);
```

Comment:
Note that there can be cycles in the flow graph even without any improper loop headers (it's the reason why I originally removed `m_improperLoopHeaders` from `FlowGraphNaturalLoops`). #96153 has some more information. I'm guessing it doesn't affect the approximation here because those cases happen only due to exceptional flow?

Reply:
Right, we assume exceptional flow is sufficiently rare that any cycle it might form can be ignored. Also we don't model flow in and out of handlers, except for normally invoked finallys.

This might need some tweaks when we run this later after the importer rewires the flow.